src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php

```diff
@@ -321,6 +333,23 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
                         $params = array_merge($params, $data[$paramName]);
                     }
                 } elseif ($allowed && !$ignored && (isset($data[$key]) || array_key_exists($key, $data))) {
+                    if ($this->propertyInfoExtractor) {
+                        $types = $this->propertyInfoExtractor->getTypes($class, $key);
+
+                        foreach ($types as $type) {
+                            if ($type && $type->getClassName() && (!empty($data[$key]) || !$type->isNullable())) {
+                                if (!$this->serializer instanceof DenormalizerInterface) {
+                                    throw new RuntimeException(sprintf('Cannot denormalize attribute "%s" because injected serializer is not a denormalizer', $key));
+                                }
+
+                                $value = $data[$paramName];
+                                $data[$paramName] = $this->serializer->denormalize($value, $type->getClassName(), $format, $context);
+
+                                break;
+                            }
+                        }
+                    }
+
```

Comment:
Maybe not entirely your responsibility, but this method is becoming very unreadable. Guard clauses could make this a lot better. E.g. everything could be moved 1 indent left by simply doing:

```
if (!$constructor) {
    return new $class();
}
```

I'm pretty sure that with some time and effort this method could become a lot less complex.